src/V4Router.sol

```diff
@@ -0,0 +1,179 @@
+/// @dev the entry point to executing actions in this contract is calling `BaseActionsRouter._executeActions`
+/// An inheriting contract should call _executeActions at the point that they wish actions to be executed
+abstract contract V4Router is IV4Router, BaseActionsRouter {
+    using CurrencyLibrary for Currency;
+    using TransientStateLibrary for IPoolManager;
+    using CalldataBytesLib for bytes;
+
+    constructor(IPoolManager poolManager) BaseActionsRouter(poolManager) {}
+
+    function _handleAction(uint256 action, bytes calldata params) internal override {
+        // swap actions and payment actions in different blocks for gas efficiency
+        if (action < 0x10) {
+            if (action == Actions.SWAP_EXACT_IN) {
+                _swapExactInput(abi.decode(params, (IV4Router.ExactInputParams)));
+            } else if (action == Actions.SWAP_EXACT_IN_SINGLE) {
+                _swapExactInputSingle(abi.decode(params, (IV4Router.ExactInputSingleParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT) {
+                _swapExactOutput(abi.decode(params, (IV4Router.ExactOutputParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT_SINGLE) {
+                _swapExactOutputSingle(abi.decode(params, (IV4Router.ExactOutputSingleParams)));
+            } else {
+                revert UnsupportedAction(action);
+            }
+        } else {
+            if (action == Actions.SETTLE) {
+                // equivalent: abi.decode(params, (Currency))
+                Currency currency;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                }
+
+                // TODO support address(this) paying too
+                _payAndSettle(currency, _msgSender());
+            } else if (action == Actions.TAKE) {
+                // equivalent: abi.decode(params, (Currency, address))
+                Currency currency;
+                address recipient;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                    recipient := calldataload(add(params.offset, 0x20))
+                }
+
+                // TODO add min amount??
+                _take(currency, recipient);
+            } else {
+                revert UnsupportedAction(action);
+            }
+        }
+    }
+
+    function _swapExactInputSingle(IV4Router.ExactInputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(-int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+
+            for (uint256 i = 0; i < pathLength; i++) {
```

Comment:
We batch here because the input to the second step is the output from the first step.
Like if you have a split-route like this:

1) 100 UNI -> DAI -> USDC
2) 50 UNI -> USDT -> USDC

How much DAI should be traded for USDC? everything that was produced by the UNI->DAI trade. So it just does this loop to pass values through. 
It also saves having so many parameters because the trade is passed in as `100 UNI->DAI->USDC` instead of `100 UNI -> DAI`, `X DAI -> USDC`.

So the example split route above would be passed in as 2 swap commands one for each route

Reply:
This is how our V2 and V3 swap routers have worked too - passing in an entire path beginning to end

Reply:
I can try to play around with passing values between commands like you suggest and see what happens

Reply:
hmmm yeah I guess this is just structurally different from POSM that doesn't do subbatching even when it "could" roll its positions. i just find it odd to write a batched actions router where the command below are basically another batch. it just means you could have 2 less commands bc you can just encode many swapExactInputSingles together to get a full swapExactInput trade

in posm, we dont have anything like this but we COULD.. ie decrease (USDC, ETH) then increase (ETH, DAI) is technically like a "path" but it requires calling decrease then increase with the full params

Reply:
yeah thats fair... i think it will cost way more gas because we will have to pass outputs between commands, or load deltas from pool manager, and pass in each token twice so way more calldata.
Its like how posm will have a command CLOSE_PAIR which could really just be CLOSE done twice, to be optimal.
I'm down to open an issue and explore it after - and we can get emily's opinion too given shes done a lot of routing work for both v4 and previous versions.

Reply:
opened #211

Reply:
sounds good, I think we may run out of time to explore this though. It makes sense as an optimization for router and the design similarities with prev routers makes sense